src/dune_pkg/opam_solver.ml

```diff
@@ -106,26 +115,123 @@ module Context_for_dune = struct
     ; candidates_cache
     ; available_cache
     ; constraints
+    ; latest_released_base_compiler_version
     }
   ;;
 
   let pp_rejection f = function
     | Unavailable -> Format.pp_print_string f "Availability condition not satisfied"
   ;;
 
+  let opam_version_compare a b = OpamPackage.Version.compare a b |> Ordering.of_int
+
+  (* Returns a bool which is true iff the following conditions are met:
+     - the version preference is [Newest]
+     - the given package is the ocaml compiler metapackage
+     - the version of the given package is later than the latest
+       version of the ocaml-base-compiler package lacking the
+       avoid-version flag *)
+  let opam_file_is_ocaml_newer_than_latest_base_compiler_if_prefer_newest
+    t
+    (opam_file : OpamFile.OPAM.t)
+    =
+    match t.version_preference with
+    | Oldest -> false
+    | Newest ->
+      (match opam_file_is_ocaml_package opam_file with
+       | false -> false
+       | true ->
+         (match Option.both opam_file.version t.latest_released_base_compiler_version with
```

Comment:
When can the versions be absent here?

Reply:
For the lhs, only if the package being compared is named "ocaml" but lacks a version, and for the rhs, only if the opam repo has no package named "ocaml-base-compiler". Neither of these cases should happen in practice but I'm trying to minimize the assumptions dune makes about the opam repo.

Though I notice that elsewhere in this file we call `OpamFile.OPAM.version` which raises if its argument doesn't have a version, so there's little point checking for the version field here.